### ISSUE TITLE
VM crash when running analyzer in LSP mode with VM Service enabled

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#46612]
I thought I'd raised this before, but I can't find it and I'm hitting this again today trying to do some CPU profiling of the analysis server.

When running the analysis server in LSP mode with `--enable-vm-service`, the VM crashes during initialisation:

```
[10:51:05 AM] [Analyzer] [Info] Spawning /Users/danny/Dev/Google/flutter/bin/cache/dart-sdk/bin/dart with args ["--enable-vm-service=9111","-DSILENT_OBSERVATORY=true","--disable-service-auth-codes","/Users/danny/Dev/Google/flutter/bin/cache/dart-sdk/bin/snapshots/analysis_server.dart.snapshot","--lsp","--client-id=VS-Code","--client-version=3.24.2"]
[10:51:05 AM] [Analyzer] [Info]     PID: 3382
[10:51:05 AM] [Analyzer] [Info] ==> Content-Length: 4884
[10:51:05 AM] [Analyzer] [Info] ==> {"jsonrpc":"2.0","id":0,"method":"initialize","params":{ // snip long VS Code initalization params
[10:51:05 AM] [Analyzer] [Info] <== The Dart DevTools debugger and profiler is available at: http://127.0.0.1:9111/devtools/#/?uri=ws%3A%2F%2F127.0.0.1%3A9111%2Fws
[10:51:06 AM] [Analyzer] [Info] <== Content-Length: 408
Content-Type: application/vscode-jsonrpc; charset=utf-8
[10:51:06 AM] [Analyzer] [Error] 
===== CRASH =====
si_signo=Segmentation fault: 11(11), si_code=1, si_addr=0x50
version=2.14.0-312.0.dev (dev) (Tue Jul 13 15:32:06 2021 -0700) on "macos_x64"
pid=3382, thread=21251, isolate_group=(nil)(0x0), isolate=(nil)(0x0)
isolate_instructions=0, vm_instructions=1066bfdc0
[10:51:06 AM] [Analyzer] [Error]   pc 0x00000001069d3a81 fp 0x00007000081981f0 dart::ServiceEvent::ServiceEvent(dart::Isolate*, dart::ServiceEvent::EventKind)+0x21
[10:51:06 AM] [Analyzer] [Error]   pc 0x00000001069c2ed5 fp 0x0000700008198340 dart::Service::SendEmbedderEvent(dart::Isolate*, char const*, char const*, unsigned char const*, long)+0x45
[10:51:06 AM] [Analyzer] [Error]   pc 0x0000000106c4c003 fp 0x0000700008198360 Dart_ServiceSendDataEvent+0xe3
[10:51:06 AM] [Analyzer] [Error]   pc 0x00000001067f13c9 fp 0x00007000081983a0 dart::bin::File::WriteFully(void const*, long long)+0xb9
[10:51:06 AM] [Analyzer] [Error]   pc 0x00000001067eddf7 fp 0x00007000081983f0 dart::bin::File::WriteFromRequest(dart::bin::CObjectArray const&)+0x2d7
[10:51:06 AM] [Analyzer] [Error]   pc 0x00000001066abf97 fp 0x0000700008198440 dart::bin::IOServiceCallback(long long, _Dart_CObject*)+0x2d7
[10:51:06 AM] [Analyzer] [Error]   pc 0x00000001068be94f fp 0x0000700008198db0 dart::NativeMessageHandler::HandleMessage(std::__1::unique_ptr<dart::Message, std::__1::default_delete<dart::Message> >)+0xbf
  pc 0x00000001068bbb03 fp 0x0000700008198e30 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x133
[10:51:06 AM] [Analyzer] [Error]   pc 0x00000001068bc129 fp 0x0000700008198e90 dart::MessageHandler::TaskCallback()+0x1e9
  pc 0x00000001069f4e68 fp 0x0000700008198f20 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x158
[10:51:06 AM] [Analyzer] [Error]   pc 0x00000001069f513f fp 0x0000700008198f50 dart::ThreadPool::Worker::Main(unsigned long)+0x6f
[10:51:06 AM] [Analyzer] [Error]   pc 0x000000010694eba5 fp 0x0000700008198fb0 dart::ThreadStart(void*)+0xb5
  pc 0x00007fff205718fc fp 0x0000700008198fd0 _pthread_start+0xe0
  pc 0x00007fff2056d443 fp 0x0000700008198ff0 thread_start+0xf
-- End of DumpStackTrace
```

To reproduce, set the following VS Code settings to ensure LSP is enabled, the VM Service is enabled and it's being logged to a file (be sure to revert them after testing):

```
{
	"dart.analyzerVmServicePort": 9111,
	"dart.analyzerLogFile": "logs/analyzer.txt",
	"dart.previewLsp": true
}
```

I was testing with a Flutter project using Flutter SDK (so using the Dart SDK included in) at hash https://github.com/flutter/flutter/commit/e990f4b9120ec7df9597a000aa678d1582f25ef1.

**Comment by DanTup** [comment]
I hit this again today, and discovered that if I use "dart language-server" the issue doesn't seem to occur, but using "dart /path/to/analysis_server.snapshot" it does. I don't know if that's just fluke or not.

Since you have to send a valid LSP package before it crashes, here's a repro script. Enter an SDK path at the top and it'll crash if `useSnapshot=true` but not if `useSnapshot=false`.

```dart
import 'dart:convert';
import 'dart:io';

/// Whether to use snapshot (if false, uses "dart languge-server" command).
const useSnapshot = true;

const sdkLocation = '/Users/danny/Dev/Dart SDKs/nightly';

const initMessage =
    '{"jsonrpc":"2.0","id":0,"method":"initialize","params":{"processId":12752,"rootUri":"file:///Users/danny/Dev/TestStuff/win32","capabilities":{}}}';


Future<void> main() async {
  final vm = '$sdkLocation/bin/dart';
  final snapshot = '$sdkLocation/bin/snapshots/analysis_server.dart.snapshot';
  final args = [
    '--enable-vm-service=0',
    '-DSILENT_OBSERVATORY=true',
    '--disable-service-auth-codes',
    '--no-serve-devtools',
    if (useSnapshot) ...[
      snapshot,
      '--lsp',
    ] else
      'language-server'
  ];

  final proc = await Process.start(vm, args);

  proc.stdout.transform(utf8.decoder).listen((s) => print('<== $s'));
  proc.stderr.transform(utf8.decoder).listen((s) => print('<== $s'));

  void write(String s) {
    print('==> $s');
    proc.stdin.add(utf8.encode(s));
  }

  write('Content-Length: ${initMessage.length}\r\n\r\n');
  write('$initMessage\r\n');

  await proc.exitCode;
}
```

```
==> Content-Length: 145
==> {"jsonrpc":"2.0","id":0,"method":"initialize","params":{"processId":12752,"rootUri":"file:///Users/danny/Dev/TestStuff/win32","capabilities":{}}}
<== Content-Length: 1773
Content-Type: application/vscode-jsonrpc; charset=utf-8
<==
===== CRASH =====
si_signo=Segmentation fault: 11(11), si_code=1, si_addr=0x58
version=2.16.0-edge.b1955d63ad678b651b09db3dd286136c4463f36b (be) (Mon Nov 22 17:02:50 2021 +0000) on "macos_x64"
pid=12988, thread=3587, isolate_group=(nil)(0x0), isolate=(nil)(0x0)
isolate_instructions=0, vm_instructions=105b309e0
  pc 0x0000000105e00033 fp 0x000070000800a6a0 dart::ServiceEvent::ServiceEvent(dart::Isolate*, dart::ServiceEvent::EventKind)+0x13
  pc 0x0000000105deff27 fp 0x000070000800a7f0 dart::Service::SendEmbedderEvent(dart::Isolate*, char const*, char const*, unsigned char const*, long)+0x37
  pc 0x00000001062732d5 fp 0x000070000800a800 Dart_ServiceSendDataEvent+0x75
  pc 0x0000000105c35fd9 fp 0x000070000800a840 dart::bin::File::WriteFully(void const*, long long)+0xa9
  pc 0x0000000105c332b5 fp 0x000070000800a890 dart::bin::File::WriteFromRequest(dart::bin::CObjectArray const&)+0x2a5
  pc 0x0000000105b1dd75 fp 0x000070000800a8e0 dart::bin::IOServiceCallback(long long, _Dart_CObject*)+0x2b5
  pc 0x0000000105cfe712 fp 0x000070000800ade0 dart::NativeMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message> >)+0x92
  pc 0x0000000105cea2a8 fp 0x000070000800ae50 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x228
  pc 0x0000000105cea8bf fp 0x000070000800aea0 dart::MessageHandler::TaskCallback()+0x1df
  pc 0x0000000105e16c08 fp 0x000070000800af30 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x138
  pc 0x0000000105e1709d fp 0x000070000800af60 dart::ThreadPool::Worker::Main(unsigned long)+0x5d
  pc 0x0000000105d89daf fp 0x000070000800afb0 dart::OSThread::GetMaxStackSize()+0xaf
  pc 0x00007fff204a98fc fp 0x000070000800afd0 _pthread_start+0xe0
  pc 0x00007fff204a5443 fp 0x000070000800aff0 thread_start+0xf
-- End of DumpStackTrace

Exited
```

**Comment by DanTup** [comment]
Oh, but it turns out it _does_ still occur when running from source (which makes it difficult for me to profile while also make changes to the code).

I'm not sure what the cause is, but if anyone knows of any possible workarounds I might be able to use in the meantime, that would be very useful.

**Comment by bkonyi** [comment]
Resolved by 9c8f51f

**Comment by DanTup** [comment]
@bkonyi awesome, thanks! :)
